src/_pytest/config/findpaths.py

```diff
@@ -24,45 +25,98 @@ def exists(path, ignore=OSError):
         return False
 
 
-def getcfg(args, config=None):
+def _parse_ini_config(path: py.path.local) -> iniconfig.IniConfig:
+    """Parses the given generic '.ini' file using legacy IniConfig parser, returning
+    the parsed object.
+
+    Raises UsageError if the file cannot be parsed.
     """
-    Search the list of arguments for a valid ini-file for pytest,
-    and return a tuple of (rootdir, inifile, cfg-dict).
+    try:
+        return iniconfig.IniConfig(path)
+    except iniconfig.ParseError as exc:
+        raise UsageError(str(exc))
+
+
+def load_config_dict_from_file(
+    filepath: py.path.local,
+) -> Optional[Dict[str, Union[str, List[str]]]]:
+    """Loads pytest configuration from the given file path, if supported.
 
-    note: config is optional and used only to issue warnings explicitly (#2891).
+    Return None if the file does not contain valid pytest configuration.
+    """
+
+    # configuration from ini files are obtained from the [pytest] section, if present.
+    if filepath.ext == ".ini":
+        iniconfig = _parse_ini_config(filepath)
+
+        if "pytest" in iniconfig:
+            return dict(iniconfig["pytest"].items())
+        else:
+            # "pytest.ini" files are always the source of configuration, even if empty
+            return {} if filepath.basename == "pytest.ini" else None
+
+    # '.cfg' files are considered if they contain a "[tool:pytest]" section
+    elif filepath.ext == ".cfg":
+        iniconfig = _parse_ini_config(filepath)
+
+        if "tool:pytest" in iniconfig.sections:
+            return dict(iniconfig["tool:pytest"].items())
+        elif "pytest" in iniconfig.sections:
+            # If a setup.cfg contains a "[pytest]" section, we raise a failure to indicate users that
+            # plain "[pytest]" sections in setup.cfg files is no longer supported (#3086).
+            fail(CFG_PYTEST_SECTION.format(filename="setup.cfg"), pytrace=False)
+        return None
+
+    # '.toml' files are considered if they contain a [tool.pytest.ini_options] table
+    elif filepath.ext == ".toml":
+        import toml
+
+        config = toml.load(filepath)
+
+        result = config.get("tool", {}).get("pytest", {}).get("ini_options", None)
+        if result is not None:
+            # TOML supports richer data types than ini files (strings, arrays, floats, ints, etc),
+            # however we need to convert all scalar values to str for compatibility with the rest
+            # of the configuration system, which expects strings only. We needed to change the
+            def make_scalar(v: Any) -> Union[str, List[str]]:
```

Comment:
```suggestion
            def make_scalar(v: object) -> Union[str, List[str]]:
```

(Use `Any` want you to be able to do anything with the value with the type checker interfering, use `object` when you want to assume nothing about the value)

Reply:
Done, thanks for the pointer. 👍